03_DeviceSelectionAndSharedSources/main.cpp

```diff
@@ -13,318 +13,422 @@ using namespace system;
+			nbl::asset::IShaderCompiler::SCompilerOptions options = {};
+			// The Shader Asset Loaders deduce the stage from the file extension,
+			// if the extension is generic (.glsl or .hlsl) the stage is unknown.
+			// But it can still be overriden from within the source with a `#pragma shader_stage` 
+			options.stage = source->getStage() == IShader::ESS_COMPUTE ? source->getStage() : IShader::ESS_VERTEX; // TODO: do smth with it
+			options.targetSpirvVersion = m_device->getPhysicalDevice()->getLimits().spirvVersion;
+			// we need to perform an unoptimized compilation with source debug info or we'll lose names of variable sin the introspection
+			options.spirvOptimizer = nullptr;
+			options.debugInfoFlags |= IShaderCompiler::E_DEBUG_INFO_FLAGS::EDIF_SOURCE_BIT;
+			// The nice thing is that when you load a shader from file, it has a correctly set `filePathHint`
+			// so it plays nicely with the preprocessor, and finds `#include`s without intervention.
+			options.preprocessorOptions.sourceIdentifier = source->getFilepathHint();
+			options.preprocessorOptions.logger = m_logger.get();
+			options.preprocessorOptions.includeFinder = compilerSet->getShaderCompiler(source->getContentType())->getDefaultIncludeFinder();
+
+			auto spirvUnspecialized = compilerSet->compileToSPIRV(source.get(), options);
+			const CSPIRVIntrospector::CStageIntrospectionData::SParams inspctParams = { .entryPoint = "main", .shader = spirvUnspecialized };
+
+			introspection = introspector.introspect(inspctParams);
+			if (!introspection)
 			{
-				IGPUQueue* const queue = getComputeQueue();
+				logFail("SPIR-V Introspection failed, probably the required SPIR-V compilation failed first!");
+				return std::pair(nullptr, nullptr);
+			}
 
-				smart_refctd_ptr<nbl::video::IGPUCommandBuffer> cmdbuf;
-				{
-					smart_refctd_ptr<nbl::video::IGPUCommandPool> cmdpool = m_device->createCommandPool(queue->getFamilyIndex(), IGPUCommandPool::ECF_TRANSIENT_BIT);
-					if (!m_device->createCommandBuffers(cmdpool.get(),IGPUCommandBuffer::EL_PRIMARY,1u,&cmdbuf))
-					{
-						logFail("Failed to create Command Buffers!\n");
-						return false;
-					}
-				
-					cmdbuf->begin(IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
-					cmdbuf->bindComputePipeline(pipeline.get());
-					cmdbuf->bindDescriptorSets(nbl::asset::EPBP_COMPUTE,pipeline->getLayout(),0,MaxDescriptorSets,&ds->get());
-					cmdbuf->dispatch(WorkgroupCount,1,1);
-					cmdbuf->end();
-				}
+			{
+				auto* srcContent = spirvUnspecialized->getContent();
 
-				// A sidenote: Waiting for the Device or Queue to become idle does not give the same memory visibility guarantees as a signalled fence or semaphore
-				smart_refctd_ptr<IGPUFence> done = m_device->createFence(IGPUFence::ECF_UNSIGNALED);
+				system::ISystem::future_t<core::smart_refctd_ptr<system::IFile>> future;
+				m_physicalDevice->getSystem()->createFile(future, system::path("../app_resources/compiled.spv"), system::IFileBase::ECF_WRITE);
```

Comment:
you probably want localCWD as the base path here

Reply:
ah you don't, please use builtin resources CMake, then you can use "app_resources/" cause app_resources will be a mountpoint